QEfficient/lora/auto.py

```diff
@@ -0,0 +1,366 @@
+            logger.warning(f"{adapter_name} has been loaded. Skip download.")
+        else:
+            if adapter_weight and adapter_config:  # if sufficiently get adapter weight and adpater config
+                self.adapter_weights[adapter_name] = adapter_weight
+                self.adapter_configs[adapter_name] = adapter_config
+            else:  # donwload with adapter_model_id
+                self.adapter_weights[adapter_name] = {
+                    k: v.numpy().astype("float16") for k, v in load_peft_weights(adapter_model_id).items()
+                }
+                self.adapter_configs[adapter_name] = PeftConfig.from_pretrained(adapter_model_id)
+
+    def load_adapter(
+        self,
+        adapter_model_id: str,
+        adapter_name: str,
+        adapter_weight: Optional[dict] = None,
+        adapter_config: Optional[PeftConfig] = None,
+    ):
+        "Load adapter into CPU cache and Sets active adapter from one of the loaded adapters"
+
+        # check if adapter name already exist and activated
+        if adapter_name in self.active_adapter_to_id.keys():
+            logger.warning(f"{adapter_name} exists and activated. Please provide a different adapter_name.")
+        else:
+            self.download_adapter(adapter_model_id, adapter_name, adapter_weight, adapter_config)
+
+            # starting from the second adapter_name, check if adapters has same target module and rank
+            if list(self.adapter_configs.values())[0] and (
+                self.adapter_configs[adapter_name].target_modules
+                != list(self.adapter_configs.values())[0].target_modules
+            ):
+                raise ValueError(
+                    f"{adapter_name} must have same target_modules as {list(self.adapter_configs.keys())[0]}"
+                )
+            if list(self.adapter_configs.values())[0] and (
+                self.adapter_configs[adapter_name].r != list(self.adapter_configs.values())[0].r
+            ):
+                raise ValueError(f"{adapter_name} must have same rank as {list(self.adapter_configs.keys())[0]}")
+
+            # set active adapter id to current max if adapter_name is new
+            if adapter_name not in self.active_adapter_to_id.keys():
+                self.active_adapter_to_id[adapter_name] = self.max_num_adapters + 1  # reserve 0 for base
+
+                # add active adapter to set
+                self.max_num_adapters = len(self.active_adapter_to_id)
+
+        return self.active_adapter_to_id[adapter_name]
+
+    def unload_adapter(self, adapter_name: str):
+        "Deactivate adpater and remove it from CPU cache"
+
+        # step1: remove from active list if it's there
+        if adapter_name not in self.active_adapter_to_id.keys():
+            logger.info(f"Adapter name {adapter_name} is not set active yet")
+            return False
+
+        self.max_num_adapters -= 1
```

Comment:
Do we really need self.max_num_adapters variable? the number of active adapters can be always derived from len(self.active_adapter_to_id)

Reply:
I have removed the `self.max_num_adapters` and use `len(self.active_adapter_to_id)` every time needed.